shared/src/main/scala/mlscript/codegen/Codegen.scala

```diff
@@ -149,6 +149,10 @@ object SourceCode {
   def concat(codes: Ls[SourceCode]): SourceCode =
     codes.foldLeft(SourceCode.empty) { _ + _ }
 
+  // concatenate source codes without intermediate allocations
+  def bulkConcat(codes: Iterable[SourceCode]): SourceCode =
+    new SourceCode(codes.map(_.lines).foldRight(List.empty[SourceLine])((lines, accum) => lines ::: accum))
```

Comment:
If you really want to avoid too many intermediate allocations:
```suggestion
    new SourceCode(codes.iterator.map(_.lines).foldRight(List.empty[SourceLine])((lines, accum) => lines ::: accum))
```

Reply:
Oh ok using an iterator does not allocate the intermediate map.